libzmqutil: add cert class and use it instead of CZMQ zcert_t

garlick opened the pull request:
This adds a `zmqutil_cert` class to replace CZMQ `zcert_t` and completes the conversion away from libczmq.

The certificate format does not change.

garlick commented:
Fixed a TOCTOU called out by asan, and a spell error caught by the spell checker.  Yay CI!

garlick commented on src/cmd/flux-keygen.c:
```diff
@@ -93,37 +117,34 @@
+    if (path) {
+        int fd;
+        FILE *f;
+        if (!(fd = open (path, O_WRONLY | O_CREAT | O_TRUNC, 0600))
```
Hmm, I'm not sure I agree with this one.  Basically it wants me to sanitize `path` before opening it for writing in `flux-keygen` but it sounds like they think it's a privileged command or something which it is not.  Any thoughts?

grondo replied:
Yeah, this seems a little silly. Will the code scanner call out every program that takes an argument to open a file? (e.g. `vi`?) Maybe @dun would have some insight, but I think this could be dismissed as a false positive.

garlick replied:
I wonder if it would be a good idea to define this cert format in a Flux RFC.  :thinking:

dun replied:
If `flux-keygen` is running as the user, I would think this would be equivalent to what they could do at the shell. So I'm not sure what the benefit would be for additional sanitization.

garlick replied:
Thanks @dun!  OK, I'll dismiss this one then.

grondo commented on src/common/libzmqutil/cert.c:
```diff
@@ -0,0 +1,396 @@
+#include "sockopt.h"
+#include "cert.h"
+
+#define KEYSIZE 32
+#define TXTSIZE 41
```
A comment of where these sizes are coming from might be helpful. I'm guessing cert keys are limited to 32 bytes (256 bit) and the max expanded size of 32 bytes with z85 is 41 bytes?

garlick commented:
Chasing an ASAN issue in the cert test and found some more stuff that needs testing along the way, so please hold off reviewing.

grondo commented on src/common/libzmqutil/test/cert.c:
```diff
@@ -0,0 +1,308 @@
+    "public-key = \".0TmVO:u.]cUa[tFi[rH<3q2Cmu?1IBq!8/Q+&(0\"\n"
+    "secret-key = \"UPk352x.N&a(0mXR&{9S^<I!cH[Oo*^BeU)UD/[5\"\n"
+    },
+};
```
Might be good to expand the "bad" tests here, including something like a "good" cert with one character change (so should fail), some garbage input, public and/or secret key too long, unclosed quote, etc

grondo commented on scripts/install-deps-rpm.sh:
```diff
@@ -8,7 +8,6 @@ yum install \
   pkgconfig \
   glibc-devel \
   zeromq4-devel \
-  czmq-devel \
```
Remove from Dockerfiles too so the next time we generate images this dependency is dropped?

grondo replied:
Or is czmq still required by some other dependency?

garlick replied:
Not that I know of but actually...maybe it makes sense to drop czmq in another PR just for release notes clarity?  This PR Is heavily about the cert class.

grondo commented on src/common/libzmqutil/cert.h:
```diff
@@ -0,0 +1,37 @@
+#include <stdio.h>
+#include <stdbool.h>
+
+struct cert *zmqutil_cert_create (void);
```
It feels a little strange that this class is `zmqutil_cert` but the structure is just `struct cert`. However, it probably doesn't rise to the level of updating the whole PR. (just a comment)

garlick replied:
Yeah and it's a narrowly used internal class so maybe just claiming the "cert_" namespace is OK here.  I can make that change.

grondo reviewed (approved):
This LGTM. Feel free to ignore suggestions inline, though adding the extra bad cert tests might be a good idea before merging.

grondo commented:
Oops sorry @garlick! I was reviewing and missed your comment to hold off!

garlick commented:
Oh that's OK,  I haven't changed that much.  Good suggestions!

garlick commented:
Just forced a push with a few changes
- rename cert functions from `zmqutil_cert_*()` to `cert_*()`
- add a check that secret and public keys "match" (I didn't realize there was `zmq_curve_public()`)
- fixed the ASAN issue with the test
- add more cert tests (including expanding the bad input as suggested)
- misc other minor cleanup
-  dropped the czmq build changes (will submit as  another PR)

garlick commented:
@grondo -  I know you already approved but maybe you could give this one more quick look to see if anything pops out?

grondo reviewed (approved):
Still LGTM!
